Adds metastore endpoints

kRITZCREEK opened the pull request:
Also renders URLs through purescript-uri instead of munging together strings

I can't figure out how to make the tests go locally, so I hope they work on Travis

Cmdv commented on src/Quasar/Advanced/QuasarAF/Interpreter/Config.purs:
```diff
@@ -17,13 +17,8 @@ limitations under the License.
-  , permissions ∷ Array TokenHash
-  | r
-  }
+type Config r = QuasarF.Config ( idToken ∷ Maybe IdToken, permissions ∷ Array TokenHash | r )
```
This is a question, what happened to `basePath`, I saw you put all this into one place from `QuasarAF/Interpreter/Aff.purs` which makes sense but then didn't understand if `basePath` was still needed 😄

kritzcreek replied:
from `Quasar.QuasarF.Interpreter.Config`:

`type Config r = { basePath ∷ BasePath | r }`

so we're just extending that here, which means we still have the basePath.

garyb commented on src/Quasar/Metastore.purs:
```diff
@@ -0,0 +1,62 @@
+  show = case _ of
+    H2Metastore s → "(H2Metastore " <> show s <> ")"
+    PostgresMetastore { host, port, database } →
+      "(PostgresMetastore { host: " <> show host <> ", port: " <> show port <> ", database: " <> show database  <> "})"
```
Could derive generic and then show for this? I've been moving towards that in here, since show instances are the worst. :smile:

garyb replied:
Oh, I guess not because of the extensible row.

garyb commented on src/Quasar/Metastore.purs:
```diff
@@ -0,0 +1,62 @@
+       ~> "password" := password
+       ~> jsonEmptyObject
+      )
+    ~> jsonEmptyObject
```
Can we use the codec stuff for this, or is it inconvenient to get the exact format without resorting to `basicCodec` anyway or something?

garyb replied:
Again, I see, it's two different codecs because of the extensible row.

garyb commented on src/Quasar/Mount/View.purs:
```diff
@@ -39,13 +40,13 @@ type Config =
 toJSON ∷ Config → Json
 toJSON config =
-  let uri = URI.printAbsoluteURI (toURI config)
+  let (uri ∷ String) = AbsoluteURI.print (toURI config)
```
Is the annotation necessary here? I thought `AbsoluteURI.print` was monomorphic.

garyb commented on src/Quasar/QuasarF.purs:
```diff
@@ -230,3 +233,11 @@ deleteMount
+putMetastore
+  ∷ { initialize ∷ Boolean, metastore ∷ Metastore (password ∷ String) }
+  → QuasarFE Unit
+putMetastore ms = PutMetastore ms id
```
We should add the appropriate versions of these in `Quasar.Advanced.QuasarAF` too.

kritzcreek commented:
I think I addressed your feedback. We might want to coordinate merging and releasing this, since it requires everything in slamdata to be compatible with purescript-uri 4.0.0.

(for example: https://github.com/slamdata/purescript-leafletjs/pull/5)

kritzcreek commented:
Btw, does this need changing with regards to https://github.com/quasar-analytics/quasar/pull/2767/files?

garyb commented:
Hmm, good point. Have you tried using it / can we add a test case for it perhaps?
